ENH: Add SIMD sin/cos implementation with numpy-simd-routines

seiko2plus opened the pull request:
numpy-simd-routines added as subrepo in meson subprojects
  directory and the current FP configuration is static, ~1ulp used for double-precision
  ~4ulp for single-precision with handling floating-point errors,
  special-cases extended precision for large arguments,
  subnormals are enabled by default too.

  numpy-simd-routines supports all SIMD extensions that are supported
  by Google Highway including non-FMA extensions and is fully independent
  from libm to guarantee unified results across all compilers and
  platforms.

  Note: that there was no SIMD optimization enabled for sin/cos
  for double-precision before, only single-precision.

# Benchmark

## X86 Platform
The following benchmark was tested against GCC 14.3.0 on an x86 CPU (Ryzen 7 7700X).

<details>
<summary>Environment</summary>

```Bash
> uname -a
Linux seiko-pc 6.12.60 #1-NixOS SMP PREEMPT_DYNAMIC Mon Dec  1 10:43:41 UTC 2025 x86_64 GNU/Linux
> gcc --version
gcc (GCC) 14.3.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> python --version
Python 3.12.12
> lscpu
Architecture:                x86_64
  CPU op-mode(s):            32-bit, 64-bit
  Address sizes:             48 bits physical, 48 bits virtual
  Byte Order:                Little Endian
CPU(s):                      16
  On-line CPU(s) list:       0-15
Vendor ID:                   AuthenticAMD
  Model name:                AMD Ryzen 7 7700X 8-Core Processor
    CPU family:              25
    Model:                   97
    Thread(s) per core:      2
    Core(s) per socket:      8
    Socket(s):               1
    Stepping:                2
    Frequency boost:         enabled
    CPU(s) scaling MHz:      72%
    CPU max MHz:             5573.0000
    CPU min MHz:             545.0000
    BogoMIPS:                8982.62
    Flags:                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl 
                             xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_leg
                             acy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp
                              ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec
                              xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_cl
                             ean flushbyasid decodeassists pausefilter pfthreshold avic vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpo
                             pcntdq rdpid overflow_recov succor smca fsrm flush_l1d amd_lbr_pmc_freeze
```
</details>

<details>
<summary>x86-64-v4</summary>

```
export NPY_DISABLE_CPU_FEATURES=""
spin bench --compare parent/main --cpu-affinity 7 -t "\(<ufunc '(cos|sin)'>"
```

| Change   | Before [b70fc774] <brings_npsr~2>   | After [c5fc8423] <brings_npsr>   |   Ratio | Benchmark (Parameter)                                                   |
|----------|-------------------------------------|----------------------------------|---------|-------------------------------------------------------------------------|
| -        | 689±6μs                             | 640±30μs                         |    0.93 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'f')        |
| -        | 683±0.9μs                           | 602±2μs                          |    0.88 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'f')        |
| -        | 682±1μs                             | 603±1μs                          |    0.88 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'f')        |
| -        | 695±0.7μs                           | 604±1μs                          |    0.87 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 2, 'f')        |
| -        | 354±0.7μs                           | 231±0.4μs                        |    0.65 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'f')        |
| -        | 363±0.2μs                           | 234±4μs                          |    0.64 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'f')        |
| -        | 1.04±0.02ms                         | 632±20μs                         |    0.61 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'f')        |
| -        | 1.02±0ms                            | 612±10μs                         |    0.6  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'f')        |
| -        | 1.32±0.01ms                         | 795±2μs                          |    0.6  | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'f') |
| -        | 1.72±0.03ms                         | 806±3μs                          |    0.47 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'f') |
| -        | 1.44±0.02ms                         | 603±0.3μs                        |    0.42 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'f') |
| -        | 5.05±0.1ms                          | 2.07±0.01ms                      |    0.41 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 2, 'd') |
| -        | 4.98±0.01ms                         | 1.88±0.03ms                      |    0.38 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 1, 'd') |
| -        | 5.95±0.02ms                         | 2.26±0.07ms                      |    0.38 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'd') |
| -        | 2.18±0.05ms                         | 792±1μs                          |    0.36 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'f') |
| -        | 1.78±0.03ms                         | 608±4μs                          |    0.34 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 1, 'f') |
| -        | 5.65±0.02ms                         | 1.87±0.2ms                       |    0.33 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'd')        |
| -        | 5.65±0.01ms                         | 1.87±0.01ms                      |    0.33 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'd')        |
| -        | 5.77±0.01ms                         | 1.90±0.01ms                      |    0.33 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'd')        |
| -        | 6.00±0.04ms                         | 1.98±0.02ms                      |    0.33 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'd') |
| -        | 1.36±0.01ms                         | 443±2μs                          |    0.32 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'f') |
| -        | 5.75±0ms                            | 1.68±0.04ms                      |    0.29 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'd')        |
| -        | 4.97±0ms                            | 1.46±0.04ms                      |    0.29 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'd') |
| -        | 2.42±0.2ms                          | 609±2μs                          |    0.25 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 2, 'f') |
| -        | 5.62±0.01ms                         | 1.25±0.02ms                      |    0.22 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 2, 'd')        |
| -        | 5.72±0ms                            | 1.25±0.01ms                      |    0.22 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'd')        |
| -        | 4.97±0ms                            | 1.08±0ms                         |    0.22 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'd') |
| -        | 7.08±0.01ms                         | 1.58±0.02ms                      |    0.22 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'd') |
| -        | 7.07±0ms                            | 1.20±0.01ms                      |    0.17 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'd') |
| -        | 1.49±0ms                            | 233±10μs                         |    0.16 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'f') |
| -        | 5.61±0ms                            | 828±0.4μs                        |    0.15 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'd')        |
| -        | 5.73±0ms                            | 859±8μs                          |    0.15 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'd')        |

</details>

<details>
<summary>x86-64-v3</summary>

```
export NPY_DISABLE_CPU_FEATURES="X86_V4"
spin bench --compare parent/main --cpu-affinity 7 -t "\(<ufunc '(cos|sin)'>"
```

| Change   | Before [b70fc774] <brings_npsr~2>   | After [c5fc8423] <brings_npsr>   |   Ratio | Benchmark (Parameter)                                                   |
|----------|-------------------------------------|----------------------------------|---------|-------------------------------------------------------------------------|
| +        | 844±1μs                             | 939±50μs                         |    1.11 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'f')        |
| -        | 4.98±0ms                            | 4.57±0.1ms                       |    0.92 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'd') |
| -        | 4.98±0ms                            | 4.20±0.01ms                      |    0.84 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'd') |
| -        | 6.03±0.06ms                         | 5.00±0.06ms                      |    0.83 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'd') |
| -        | 6.06±0.06ms                         | 4.78±0.04ms                      |    0.79 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'd') |
| -        | 1.37±0.01ms                         | 913±40μs                         |    0.67 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'f')        |
| -        | 1.37±0.01ms                         | 901±10μs                         |    0.66 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'f')        |
| -        | 1.45±0ms                            | 941±2μs                          |    0.65 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'f')        |
| -        | 830±0.3μs                           | 511±2μs                          |    0.62 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'f')        |
| -        | 7.07±0.01ms                         | 4.35±0.1ms                       |    0.61 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'd') |
| -        | 844±2μs                             | 496±5μs                          |    0.59 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'f')        |
| -        | 1.46±0.01ms                         | 864±9μs                          |    0.59 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'f')        |
| -        | 3.21±0.03ms                         | 1.83±0ms                         |    0.57 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'f') |
| -        | 7.08±0.01ms                         | 3.98±0.06ms                      |    0.56 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'd') |
| -        | 3.63±0.09ms                         | 1.84±0.01ms                      |    0.51 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'f') |
| -        | 5.65±0.01ms                         | 2.82±0.1ms                       |    0.5  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'd')        |
| -        | 5.64±0ms                            | 2.74±0.3ms                       |    0.49 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'd')        |
| -        | 5.75±0ms                            | 2.76±0.1ms                       |    0.48 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'd')        |
| -        | 3.00±0.09ms                         | 1.44±0ms                         |    0.48 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'f') |
| -        | 3.78±0.1ms                          | 1.82±0.02ms                      |    0.48 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'f') |
| -        | 5.77±0.01ms                         | 2.68±0.02ms                      |    0.46 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'd')        |
| -        | 5.73±0ms                            | 2.11±0.03ms                      |    0.37 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'd')        |
| -        | 5.62±0ms                            | 2.03±0.01ms                      |    0.36 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 2, 'd')        |
| -        | 5.62±0.02ms                         | 1.67±0ms                         |    0.3  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'd')        |
| -        | 5.72±0ms                            | 1.73±0ms                         |    0.3  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'd')        |
| -        | 3.23±0.02ms                         | 866±10μs                         |    0.27 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'f') |
| -        | 3.48±0.1ms                          | 856±3μs                          |    0.25 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 2, 'f') |
| -        | 3.68±0.06ms                         | 851±2μs                          |    0.23 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 1, 'f') |
| -        | 3.03±0.08ms                         | 497±4μs                          |    0.16 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'f') |

</details>

<details>
<summary>x86-64-v2</summary>

```
export NPY_DISABLE_CPU_FEATURES="X86_V4 X86_V3"
spin bench --compare parent/main --cpu-affinity 7 -t "\(<ufunc '(cos|sin)'>"
```

| Change   | Before [b70fc774] <brings_npsr~2>   | After [c5fc8423] <brings_npsr>   |   Ratio | Benchmark (Parameter)                                                   |
|----------|-------------------------------------|----------------------------------|---------|-------------------------------------------------------------------------|
| +        | 4.98±0ms                            | 7.04±0.1ms                       |    1.41 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 2, 'd') |
| +        | 2.12±0ms                            | 2.88±0.05ms                      |    1.36 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'f') |
| +        | 4.97±0.01ms                         | 6.68±0.02ms                      |    1.34 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 1, 'd') |
| +        | 2.12±0ms                            | 2.84±0.01ms                      |    1.34 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'f') |
| +        | 2.13±0ms                            | 2.82±0ms                         |    1.32 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'f') |
| +        | 4.98±0.01ms                         | 6.19±0.05ms                      |    1.24 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'd') |
| +        | 4.97±0ms                            | 5.77±0.05ms                      |    1.16 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'd') |
| +        | 5.96±0.03ms                         | 6.90±0.2ms                       |    1.16 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 2, 'd') |
| +        | 2.13±0ms                            | 2.45±0ms                         |    1.15 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'f') |
| +        | 5.95±0.06ms                         | 6.75±0.03ms                      |    1.13 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 4, 1, 'd') |
| -        | 7.07±0.01ms                         | 6.31±0.02ms                      |    0.89 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 2, 'd') |
| -        | 7.08±0ms                            | 5.86±0.03ms                      |    0.83 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'sin'>, 1, 1, 'd') |
| -        | 5.64±0.01ms                         | 4.47±0.3ms                       |    0.79 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'd')        |
| -        | 5.76±0.01ms                         | 4.15±0.1ms                       |    0.72 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'd')        |
| -        | 5.64±0.02ms                         | 3.94±0.06ms                      |    0.7  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'd')        |
| -        | 5.76±0.01ms                         | 4.01±0.1ms                       |    0.7  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'd')        |
| -        | 2.01±0.01ms                         | 1.36±0.02ms                      |    0.67 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 1, 'f') |
| -        | 2.02±0.02ms                         | 1.35±0.03ms                      |    0.67 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 4, 2, 'f') |
| -        | 2.02±0ms                            | 1.30±0ms                         |    0.64 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 2, 'f') |
| -        | 5.62±0.01ms                         | 3.54±0.08ms                      |    0.63 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 2, 'd')        |
| -        | 5.73±0.01ms                         | 3.58±0.06ms                      |    0.62 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'd')        |
| -        | 5.62±0.01ms                         | 3.11±0.07ms                      |    0.55 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'd')        |
| -        | 5.72±0.01ms                         | 3.12±0.05ms                      |    0.54 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'd')        |
| -        | 2.01±0ms                            | 926±3μs                          |    0.46 | bench_ufunc_strides.UnaryFPSpecial.time_unary(<ufunc 'cos'>, 1, 1, 'f') |
| -        | 3.29±0ms                            | 1.45±0.1ms                       |    0.44 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 2, 'f')        |
| -        | 3.31±0ms                            | 1.41±0.04ms                      |    0.42 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 2, 'f')        |
| -        | 3.31±0ms                            | 1.36±0.01ms                      |    0.41 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 4, 1, 'f')        |
| -        | 3.28±0ms                            | 1.30±0ms                         |    0.4  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 2, 'f')        |
| -        | 3.29±0ms                            | 1.32±0.03ms                      |    0.4  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 4, 1, 'f')        |
| -        | 3.30±0ms                            | 1.32±0.03ms                      |    0.4  | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 2, 'f')        |
| -        | 3.30±0ms                            | 947±4μs                          |    0.29 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'sin'>, 1, 1, 'f')        |
| -        | 3.28±0ms                            | 925±2μs                          |    0.28 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'f')        |

</details>

#### Binary size change: _multiarray_umath.cpython-312-x86_64-linux-gnu.so

|  stripped |  size (bytes) |
|------------|--------------|
| **before** | 9,895,936    |
| **after**  | 10,014,720   |
| **diff**   | **+118,784** |

seiko2plus committed:
ENH: Add SIMD sin/cos implementation with numpy-simd-routines

  numpy-simd-routines added as subrepo in meson subprojects
  directory and the current FP configuration is static, ~1ulp used for double-precision
  ~4ulp for single-precision with handling floating-point errors,
  special-cases extended precision for large arguments,
  subnormals are enabled by default too.

  numpy-simd-routines supports all SIMD extensions that are supported
  by Google Highway including non-FMA extensions and is fully independent
  from libm to guarantee unified results across all compilers and
  platforms.

  Full benchmarks will be provided within the pull-request, the following
  benchmark was tested against clang-19 and x86 CPU (Ryzen7 7700X)
  with AVX512 enabled.

  Note: that there was no SIMD optimization enabled for sin/cos
  for double-precision, only single-precision.

  | Before        | After       |  Ratio | Benchmark (Parameter)                    |
  |---------------|-------------|--------|------------------------------------------|
  | 713±6μs       | 633±6μs     |   0.89 | UnaryFP(<ufunc 'cos'>, 1, 2, 'f')        |
  | 717±9μs       | 637±6μs     |   0.89 | UnaryFP(<ufunc 'cos'>, 4, 1, 'f')        |
  | 705±3μs       | 607±10μs    |   0.86 | UnaryFP(<ufunc 'sin'>, 4, 1, 'f')        |
  | 714±10μs      | 595±0.5μs   |   0.83 | UnaryFP(<ufunc 'sin'>, 1, 2, 'f')        |
  | 370±0.3μs     | 277±4μs     |   0.75 | UnaryFP(<ufunc 'cos'>, 1, 1, 'f')        |
  | 373±2μs       | 236±0.6μs   |   0.63 | UnaryFP(<ufunc 'sin'>, 1, 1, 'f')        |
  | 1.06±0.01ms   | 648±3μs     |   0.61 | UnaryFP(<ufunc 'cos'>, 4, 2, 'f')        |
  | 1.06±0.01ms   | 617±30μs    |   0.58 | UnaryFP(<ufunc 'sin'>, 4, 2, 'f')        |
  | 5.06±0.06ms   | 2.61±0.3ms  |   0.52 | UnaryFPSpecial(<ufunc 'cos'>, 4, 2, 'd') |
  | 1.48±0ms      | 715±5μs     |   0.48 | UnaryFPSpecial(<ufunc 'sin'>, 1, 2, 'f') |
  | 1.50±0.01ms   | 639±6μs     |   0.43 | UnaryFPSpecial(<ufunc 'cos'>, 1, 2, 'f') |
  | 5.15±0.1ms    | 1.96±0.01ms |   0.38 | UnaryFPSpecial(<ufunc 'cos'>, 4, 1, 'd') |
  | 5.72±0.02ms   | 2.09±0.1ms  |   0.37 | UnaryFP(<ufunc 'cos'>, 4, 2, 'd')        |
  | 5.76±0.01ms   | 2.03±0.08ms |   0.35 | UnaryFP(<ufunc 'sin'>, 4, 2, 'd')        |
  | 5.07±0.08ms   | 1.76±0.2ms  |   0.35 | UnaryFPSpecial(<ufunc 'cos'>, 1, 2, 'd') |
  | 6.04±0.04ms   | 2.05±0.09ms |   0.34 | UnaryFPSpecial(<ufunc 'sin'>, 4, 2, 'd') |
  | 5.79±0.03ms   | 1.90±0.2ms  |   0.33 | UnaryFP(<ufunc 'sin'>, 4, 1, 'd')        |
  | 2.29±0.1ms    | 762±40μs    |   0.33 | UnaryFPSpecial(<ufunc 'sin'>, 4, 1, 'f') |
  | 5.72±0.1ms    | 1.75±0.07ms |   0.31 | UnaryFP(<ufunc 'cos'>, 4, 1, 'd')        |
  | 6.04±0.03ms   | 1.82±0.2ms  |   0.3  | UnaryFPSpecial(<ufunc 'sin'>, 4, 1, 'd') |
  | 2.49±0.1ms    | 748±30μs    |   0.3  | UnaryFPSpecial(<ufunc 'sin'>, 4, 2, 'f') |
  | 2.23±0.1ms    | 634±6μs     |   0.28 | UnaryFPSpecial(<ufunc 'cos'>, 4, 1, 'f') |
  | 1.31±0.03ms   | 367±5μs     |   0.28 | UnaryFPSpecial(<ufunc 'sin'>, 1, 1, 'f') |
  | 2.55±0.09ms   | 654±30μs    |   0.26 | UnaryFPSpecial(<ufunc 'cos'>, 4, 2, 'f') |
  | 4.97±0.03ms   | 1.14±0ms    |   0.23 | UnaryFPSpecial(<ufunc 'cos'>, 1, 1, 'd') |
  | 5.67±0.01ms   | 1.22±0.03ms |   0.22 | UnaryFP(<ufunc 'cos'>, 1, 2, 'd')        |
  | 5.76±0.03ms   | 1.28±0.06ms |   0.22 | UnaryFP(<ufunc 'sin'>, 1, 2, 'd')        |
  | 1.26±0.01ms   | 272±2μs     |   0.22 | UnaryFPSpecial(<ufunc 'cos'>, 1, 1, 'f') |
  | 7.03±0.02ms   | 1.31±0.01ms |   0.19 | UnaryFPSpecial(<ufunc 'sin'>, 1, 2, 'd') |
  | 5.67±0.01ms   | 810±9μs     |   0.14 | UnaryFP(<ufunc 'cos'>, 1, 1, 'd')        |
  | 5.71±0.01ms   | 817±40μs    |   0.14 | UnaryFP(<ufunc 'sin'>, 1, 1, 'd')        |
  | 7.05±0.03ms   | 915±4μs     |   0.13 | UnaryFPSpecial(<ufunc 'sin'>, 1, 1, 'd') |

seiko2plus committed:
Relax sin/cos ULP test for float32 on non-FMA

Allow up to 3 ULP error for float32 sin/cos when native
FMA is not available.

rgommers commented:
> At the moment, I only have access to bare-metal x86 hardware; Perhaps I should try cloud services—e.g., AWS—for testing on other architectures. 

No more GCC compile farm access? AWS seems okay of course. If you want me to run this on macOS arm64 (M1), I can.

> The ufunc inner implementation is optimized for size as much as possible and only increases +118,784B (stripped) on x86/gcc.

That's impressively small, glad to see that.

seberg commented:
@seiko2plus thanks a lot for this hard work!  I don't want to discuss it on the PR here, but maybe you can send a very brief mail summarizing precision changes for the float32 version? (I am slightly worried might get into the same dance as with the 64bit version and SVML before, so should at least increase awareness.)

mhvk reviewed (commented):
@seiko2plus - looks very impressive! A few small comments inline, with that about the iterator perhaps the most important (if it is possible; can be for follow-up). I also made a small comment over at your new npsr routines.

mhvk commented on numpy/_core/src/umath/loops_trigonometric.dispatch.cpp:
```diff
@@ -1,299 +1,119 @@
+    // Note: the intrinsics of NumPy SIMD routines are inlined by default.
+    if (NPY_UNLIKELY(is_mem_overlap(args[0], sin, args[1], sout, len) ||
+                     sin != sizeof(T) || sout != sizeof(T))) {
+        // this for non-contiguous or overlapping case
```
@seberg - you mentioned there is no flag, but is one needed? If `sin`/`cos` are defined as an array method, would things not work automatically if one filled only the `contiguous_loop` slot? Or is that at the moment only treated as something that will be used if present if data are contiguous.

mhvk replied:
> > I would think the overhead cannot be that much worse than it is here
> 
> Well, the buffering has a "huge" amount of one time overhead. For very large arrays, yes the overhead may well be lower in practice (or at least nicer, as it's repeated fewer times).

The other thing is that if we centralize this, there will be more of an incentive to speed it up.  Isn't the buffer pre-allocated by the compiler here? It would make more sense to have just one scratch buffer in the iterator, perhaps using your mechanism that allocates if it is too small.

seberg replied:
@mhvk no not yet, the contiguous loop is an optimization only, only with that flag or a "non-contig to contig wrapper" could it be more. Dunno what is better, centralizing or not, I mostly like it because it removes unnecessary repitition both in code and binary, whether it is slower or faster.
(but yeah, we shouldn't worry about that here.)

seiko2plus committed:
SIMD, TST: Apply Marten’s suggestions

- Extend the C++ doc scope to better explain precision control, which is chosen based on the data type.
- Add test cases for sine/cosine facts—for example, `cos(0) == 1`.

seiko2plus committed:
SIMD: Enable SVE/RVV dynamic dispatch for sin/cos kernels

seberg commented:
I don't really remember the state of this PR (it is tagged for 2.5).  Should we push for it?

One serious worry I have is still this part:
> ~4ulp for single-precision with handling floating-point errors,
which to me seems very optimistic that nobody cares (and we had discussed it also in some meeting to a similar result). Yes many (maybe most) float32 users may not care about precision, but that doesn't mean that no-one will be bitten by it unfortunately.
My opinion is that we really need a "fast-math" kind of switch for this type of thing, and otherwise stay roughly within e.g. glibc precision.

Mousius commented:
@seberg I think the *current* FP32 routines are also 4ULP, so this doesn't change anything and we can move forwards with it. I'm in agreement with you that we should give users the ability to opt-in/out of lower precision though.

seberg commented on numpy/_core/tests/test_umath.py:
```diff
@@ -2202,19 +2251,30 @@ def test_sincos_float32(self):
             np.sin(x_f32, out=x_f32),
             np.float32(np.sin(x_f64)),
-            maxulp=2,
+            maxulp=maxulp_f32,
```
@Mousius if there is no big change then we don't have to worry about it, but this looks a bit like there may be one?  (But maybe not, fma not being available sounds possibly niche these days, but not sure.)

Mousius replied:
My assumption with this is that the expectation was always 4ULP, but some of the routines would perform better than that.

seiko2plus replied:
`maxulp_f32` is `~3` ulp for SIMD extensions without native FMA support, which I think is acceptable. If we wanted to tighten that, we could add an extra correction step — but I'd prefer to do that only after we have a proper test suite in place (we need a large dataset to validate edge cases reliably).

As a quick workaround, if the relaxed tolerance is really a concern, we can temporarily force the high-precision profile for those extensions until the correction step is implemented.

seiko2plus replied:
Ah, also I suppose we agreed that the low-prec profile can even go further to `~4ulp`.

seberg replied:
> Ah, also I suppose we agreed that the low-prec profile can even go further to ~4ulp.

This was the reason why I commented: I am not sure that this was ever quite agreed upon. And if it was, I suspect that this was before the recent fallout and discussion around lower fp16 precision.
https://mail.python.org/archives/list/numpy-discussion@python.org/thread/QLR57SMUPQVMGKXWQYR5CSE6JBPK43IG/#4GWQFDFYG7HCBHQEX5K4VP7DDGN25ZNZ

Having noticeably lower precision than typical math library/current code seems like something that we need to be very careful about to me.

Mousius replied:
Yeash, the conversation around making it all toggle-able was a long time ago. 

Just to clarify - do you want to block this PR @seberg? I don't see it as a change in behaviour.

seberg replied:
I commented on the PR mainly based on the PR description.  We had serious issues with float64 being less precise, and whatever I discuss/think about it, I am not convinced we can just say that float32 you can cut corners without worry.  The same argument was made with float16 and we had to revert it.

So blocking this PR?  No, because you are both promising that there is in practice no precision change.  But blocking anything that makes precision worse and argues it is OK because 4 ULP are OK for float32: yes.

After the whole float16 thing *and* the float64 precision change issue and with people noticing different precision across architectures, I have serious concerns about saying that being a bit less precise than normal C-code/glibc code is OK.
(That isn't to say that we have to match it _exactly_, but I have always been nervous about going noticeably less precise and I feel like we have been bitten quite often when being optimistic about it.)

Or maybe in other words: I am not happy to keep being optimistic about this, if we really want a much faster but less precise version, I think we should go through the trouble of implementing the choice.

seiko2plus commented:
My apologies for the delayed response.

@rgommers,

> No more GCC compile farm access? AWS seems okay of course. If you want me to run this on macOS arm64 (M1), I can.

My concerns were mainly for `SVE`, which is not supported by `M1`. I was expecting a performance gain similar to `AVX2`, but we need to dig more, I suppose. However, the current gain is acceptable. It seems AWS is the only option I have.

@seberg,

> thanks a lot for this hard work! I don't want to discuss it on the PR here, but maybe you can send a very brief mail summarizing precision changes for the float32 version? (I am slightly worried might get into the same dance as with the 64bit version and SVML before, so should at least increase awareness.)

I sent an email a long time ago:
https://mail.python.org/archives/list/numpy-discussion@python.org/thread/YIU5TZOENCRQ7GUZRKXYMUK2EIYWJINE/
And I have clarified the default precision of ~4ulp for `fp32` and ~1ulp for `fp64`.

One thing worth noting here: we already ship an enabled SIMD kernel for `float32`, and it is largely equivalent to the low-precision profile in this PR. The two differences are that the existing kernel falls back to `libm` for large inputs and lacks support for non-FMA SIMD extensions.


> I don't really remember the state  of this PR (it is tagged for 2.5). Should we push for it?

The build was totally broken due to the `meson` subproject functionality not working as it's supposed to due to a diverged rebase, which is resolved by numpy/meson#27, #31633.

New CI errors are caused by the LUT optimization patch numpy/numpy-simd-routines#7; (my bad) I will prepare a fix for it.


>  looks very impressive! A few small comments inline, with that about the iterator perhaps the most important (if it is possible; can be for follow-up). I also made a small comment over at your new npsr routines.

@mhvk,
Thank you, I have covered your suggestions in a separated commit:
 https://github.com/numpy/numpy/pull/29699/commits/c981beb27b26d51e913992edaeb235ac5285c447

@Mousius,
> Generally seems better overall, testing on some metal AWS instances:
> | -        | 5.84±0.01ms                | 3.96±0ms            |    0.68 | bench_ufunc_strides.UnaryFP.time_unary(<ufunc 'cos'>, 1, 1, 'd')                  |

Not good to me or at least not as I was expecting. Just wondering, I will dig into it. We should have a similar gain to `X86_V3`, but definitely not within this PR.

@mhvk,
> Looks like the only slowdowns are UnaryFPSpecial for double -- is that because the data in that case include "weird" numbers (like denormal). If so, this indeed seems a great improvement.

`inf`/`nans` too, the performance gain is not good overall (compared to x86) even for the happy paths, but it's acceptable for now.

seiko2plus commented on numpy/_core/src/common/simd/simd.hpp:
```diff
@@ -67,6 +68,37 @@ namespace hn = hwy::HWY_NAMESPACE;
+/// npsr is tag free by design so we only include it within main namespace (np::simd)
+namespace sr = npsr::HWY_NAMESPACE;
+
+/// Default precision configrations for NumPy SIMD Routines
```
done
